### PR TITLE
feat(cubestore): Aggregating index in pre-aggregations

### PR DESCRIPTION
**Pull request by waralexrom**
**Check List**
- [ ] Tests has been run in packages where changes made if available
- [ ] Linter has been run for changed code
- [ ] Tests for the changes have been added if not covered yet
- [ ] Docs have been added / updated if required

**Description of Changes Made (if issue reference is not provided)**

Added the `type` field to the index description in the schema. This field can have two values - `regular` (default) and `aggregate`.
I also added the corresponding changes to the `CREATE TABLE` queries.

### CODE REVIEW

**Review by cristipp — COMMENTED** [review]
There are no tests. Would be nice to have some tests, at least wrt. `CubeStoreDriver` sql generation. IMHO we should separate `sql generation` from `issue sql query` in the CubeStoreDriver implementation, so we can test it via unit-tests. In my experience, `sql code gen` can be fragile and error-prone.

File: packages/cubejs-query-orchestrator/src/driver/driver.interface.ts [hunk 1]
```diff
@@ -97,7 +103,7 @@ export type DownloadQueryResultsResult = DownloadQueryResultsBase & (DownloadTab
 export interface DriverInterface {
   createSchemaIfNotExists(schemaName: string): Promise<any>;
   uploadTableWithIndexes(
-    table: string, columns: TableStructure, tableData: DownloadTableData, indexesSql: IndexesSQL, uniqueKeyColumns: string[], queryTracingObj: any
```
[the remark this answers is not in the document]
Comment (waralexrom) [reply packages/cubejs-query-orchestrator/src/driver/driver.interface.ts:100]: Historically `IndexesSQL` is typical indexes in form of plain string with `CREATE INDEX` statement. This SQLs can be used by any driver. `createTableIndexes` is the structure used by CubeStoreDriver for place indexes of different types (regular, aggregate and etc) directly into `CREATE TABLE` statement. I.e. `CREATE TABLE ..... INDEX ... (), AGGREGATE INDEX ... (...)`
So `IndexesSQL` not used by CubeStoreDriver at all

File: packages/cubejs-schema-compiler/src/adapter/PreAggregations.js [hunk 2]
```diff
@@ -166,22 +189,36 @@ export class PreAggregations {
...
+                cube,
+                preAggregation,
+                preAggregation.indexes[index],
+                indexName,
+                tableName
+              )
+            };
+          }
+        ),
+      createTableIndexes: Object.keys(preAggregation.indexes || {})
```
[the remark this answers is not in the document]
Comment (waralexrom) [reply packages/cubejs-schema-compiler/src/adapter/PreAggregations.js:210]: No, it shouldn't, see comment above about `indexesSql` and `createTableIndexes`

File: packages/cubejs-schema-compiler/src/adapter/PreAggregations.js [hunk 3]
```diff
@@ -118,6 +118,26 @@ export class PreAggregations {
     return this.hasCumulativeMeasuresValue;
   }
 
+  aggregatesColumns(cube, preAggregation) {
```
[the remark this answers is not in the document]
Comment (waralexrom) [reply packages/cubejs-schema-compiler/src/adapter/PreAggregations.js]: done

File: packages/cubejs-schema-compiler/src/compiler/CubeValidator.js [hunk 4]
```diff
@@ -161,7 +161,8 @@ const BasePreAggregationWithoutPartitionGranularity = {
       sql: Joi.func().required()
     }),
     Joi.object().keys({
-      columns: Joi.func().required()
+      columns: Joi.func().required(),
+      type: Joi.any().valid('regular', 'aggregate').required().default('regular'),
```
[the remark this answers is not in the document]
Comment (waralexrom) [reply packages/cubejs-schema-compiler/src/compiler/CubeValidator.js]: fixed

File: packages/cubejs-cubestore-driver/src/CubeStoreDriver.ts [hunk 5]
```diff
@@ -100,30 +100,41 @@ export class CubeStoreDriver extends BaseDriver implements DriverInterface {
...
-      await this.importRows(table, columns, indexesSql, tableData, queryTracingObj);
+      await this.importRows(table, columns, indexes, aggregations, tableData, queryTracingObj);
     } else {
       throw new Error(`Unsupported table data passed to ${this.constructor}`);
     }
   }
 
-  private async importRows(table: string, columns: Column[], indexesSql: any, tableData: DownloadTableMemoryData, queryTracingObj?: any) {
-    await this.createTable(table, columns);
+  private createIndexString(index: CreateTableIndex) {
```
[the remark this answers is not in the document]
Comment (waralexrom) [reply packages/cubejs-cubestore-driver/src/CubeStoreDriver.ts:181]: I added aggregating index to this logic of `CreateTableOptions`

File: packages/cubejs-cubestore-driver/src/CubeStoreDriver.ts [hunk 6]
```diff
@@ -180,7 +191,7 @@ export class CubeStoreDriver extends BaseDriver implements DriverInterface {
 
       const createTableSql = this.createTableSql(table, columns);
       // eslint-disable-next-line no-unused-vars
-      const createTableSqlWithoutLocation = `${createTableSql}${indexes ? ` ${indexes}` : ''}`;
+      const createTableSqlWithoutLocation = `${createTableSql}${aggregations}${indexes ? ` ${indexes}` : ''}`;
```
[the remark this answers is not in the document]
Comment (waralexrom) [reply packages/cubejs-cubestore-driver/src/CubeStoreDriver.ts]: It fixed with using `createTableWithOptions` instead of this line

**Review by paveltiunov — APPROVED** [review]

**Pull request merged**